### ISSUE TITLE
Selecting .venv/bin/python as an active interpreter in VSCode doesn't enable features

### ISSUE THREAD

**Issue by oulianov** [astral-sh/uv#9637]
Hello, 

I develop my Python app with uv to handle dependencies. I want to benefit from intellisense, type hints, mypy and ruff in VSCode while I develop my app. VSCode is the most popular IDE for Python developers.

Usually, I create the a virtual env with `python -m venv`. I then select the virtual env in VSCode as an "Active Interpreter". This adds support from all the VSCode features mentioned. 

With uv, a folder .venv is created. However, when I select this .venv as an active interpreter in VSCode, **nothing changes in my editor.** Intellisense, type hints, and mypy don't work. Installed modules are not found. Custom modules are not found either. 

When I run "Reload Window", this results in a toast on the bottom right "Reactivating terminals" that never completes.

Restarting VSCode with the config doesn't make it work. 

This means for now I need to have multiple virtual environment running. One for uv, another for development. This voids the usefulness of uv for package management. 

**How to make uv virtual environments work with VSCode?**

**Comment by oulianov** [comment]
Hello, Zanieb!

OK, this is a relief to know this is supposed to work. Here are more details:


Following the same steps as you, this is what I get after running `uv add httpx` in VSCode : "Can't resolve httpx import". 
This is ok : I need to activate the python environment.
![Image](https://github.com/user-attachments/assets/23443f44-2c84-4972-8c65-ca7b8462c8ec)

I do Command+Shift+P. This lists the interpreters, but the local .venv doesn't show up. This is alright, I can still enter the path to it manually. 
![Image](https://github.com/user-attachments/assets/5da66123-877a-419e-926a-223cd966bea1)

Then, I manually select the path to the .venv created by uv: `.venv/bin/python` by clicking on "Enter Interpreter path". 
This doesn't enable intellisense.  
I can select any other environment (either conda, either created with python -m venv) normally. 
A message saying "Reactivating terminals" in the bottom left runs infinitely. Reloading the window doesn't stop this. I would need to quit and restart vscode. 
![Image](https://github.com/user-attachments/assets/225d868a-6b2b-4fe4-83be-a3da959fb529)

If I try to do `uv run hello.py` in the VSCode terminal, the process hangs and never completes. If I Ctrl+C to stop the execution and instead do `python hello.py`, the program runs normally. 

Running `uv run hello.py` in a new terminal results in this:
```
error: Querying Python at `/Users/nicolasoulianov/test/example/.venv/bin/python3` failed with exit status signal: 9 (SIGKILL)
--- stdout:

--- stderr:

---
```
![Image](https://github.com/user-attachments/assets/7d56ef8e-9304-4f10-909d-571932d8adb7)


When I go again to select an interpreter, VSCode tells me that I'm using `~/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python`.
![Image](https://github.com/user-attachments/assets/8b5422ab-cb1a-4931-a08e-227e8ad0049b)


- I have conda `23.7.4` installed. I'm using MacOS Sonoma 14.6.1 on a MacBook M3 Pro. 
- A colleague has the same problem (he use a similar machine to mine).

What else can I share with you to help you further?

**Comment by oulianov** [comment]
I think I figured this out: the uv .venv can't be selected in VSCode when the .venv is not at the root of the VSCode workspace. 

For example, this doesn't work: .venv is in example/.venv
![Image](https://github.com/user-attachments/assets/17c9adfa-fe8d-4dc9-bae4-23568a9d0292)

However, this works: .venv is at the root of the VSCode workspace.
![Image](https://github.com/user-attachments/assets/a54397a7-2196-4340-be88-1a20473dd9ca)

This is pretty weird. I'm working with a lot of monorepo so I'm used to have multiple Python projects with different configs in the same VSCode window. 

Do you know of a way to circumvent this behavior ? So that I can still have intellisense, even if the uv `.venv` is not at the root of the VSCode workspace ?

**Comment by zanieb** [comment]
That's.. very weird. I'm not sure what would cause that. 

Perhaps we should narrow down `uv run python` failing? Can you produce that consistently? 

Can you invoke Python directly? e.g.

```
❯ uv python find --python-preference only-managed --system
/Users/zb/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/bin/python3.12
❯ /Users/zb/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/bin/python3.12 --version
Python 3.12.7
```

Can you create a virtual environment with it and use it?

```
❯ /Users/zb/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/bin/python3.12 -m venv ./venv
❯ ./venv/bin/python --version
Python 3.12.7
```

**Comment by oulianov** [comment]
Unfortunately, I can't reproduce the issue with just `uv run python` 

I can invoke Python directly. I can create venv with it and use it.

```
nicolasoulianov@MacBook-Pro-de-Nicolas-2 teleop % uv python find --python-preference only-managed --system
/Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10
nicolasoulianov@MacBook-Pro-de-Nicolas-2 teleop % /Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 --version
Python 3.10.15
nicolasoulianov@MacBook-Pro-de-Nicolas-2 teleop % /Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 -m venv ./venv2
nicolasoulianov@MacBook-Pro-de-Nicolas-2 teleop % ./venv2/bin/python --version
Python 3.10.15
```

Here is another piece of information. When I select the interpreter in .venv/bin/python, I see that I select a shortcut towards the true interpreter (localized at `/Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10`). Maybe this information is not enough for VSCode to make sense of the dependencies? 
![Image](https://github.com/user-attachments/assets/ff2df880-d1a1-4ba8-8e5f-b225c049c9d4)


Do you work with uv-managed .venv in subfolders in VSCode ? Or you know someone that does ?
How does Intellisense work for you and them ?

**Comment by oulianov** [comment]
I opened a new VSCode window and tried again activating the .venv environment. Now, when I run

```
/Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 --version
```

The program hangs. I can't even Ctrl+C to stop it. 

![Image](https://github.com/user-attachments/assets/2493728b-5507-4242-b7ce-98bdfb76a00e)

I have the infinite "Reactivating terminals" loading bar
![Image](https://github.com/user-attachments/assets/159bf625-41fa-4dc1-8533-4258661421ca)


I need to restart my computer to be able to use everything normally again.

**Comment by zanieb** [comment]
Does everything work if you do `uv venv --python-preference only-system` to skip using our managed interpreters?

Do you have an antivirus or similar?

Can you look at the VSCode Python debug logs?

**Comment by oulianov** [comment]
I enabled VSCode Python debug logs. 

I don't have an antivirus. However, I'm running Vanta Agent and 1Password. 

## Using managed venv

If I select the managed uv .venv/bin/python as an active interpreter, I have a null error code. 
```
2024-12-05 16:37:48.507 [info] > ~/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 -I ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py
2024-12-05 16:37:48.511 [error] Error: Command failed: /Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 -I /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py

    at genericNodeError (node:internal/errors:984:15)
    at wrappedFn (node:internal/errors:538:14)
    at ChildProcess.exithandler (node:child_process:423:12)
    at ChildProcess.emit (node:events:531:35)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:519:28)
    at Pipe.<anonymous> (node:net:339:12) {
  code: null,
  killed: false,
  signal: 'SIGKILL',
  cmd: '/Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 -I /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py'
}
2024-12-05 16:37:50.513 [info] > ~/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10 -I ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py
```

I have then the "Reactivating terminals error"

Running 
```
/Users/nicolasoulianov/.local/share/uv/python/cpython-3.10.15-macos-aarch64-none/bin/python3.10
```
In a terminal hangs the process. It doesn't show the Python console.


## Creating a new venv

I create a .venv with `uv venv --python-preference only-system` to skip the managed interpreters.

I try to select it but nothing happens, not even an error message. I delete the conda environment I had which was named `teleop`, the same name as the folder containing the .venv. 

I try again to select the .venv. Now, I have an error when selecting it. The error code is 126. I see "Permission denied". I also have a visible error toaster in VSCode. This is the first time it has been the case. 

![Image](https://github.com/user-attachments/assets/fdb7c8e1-26c2-471d-8965-535ab0810291)

```
2024-12-05 16:50:38.156 [error] Error: Command failed: /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python -I /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py
/bin/sh: /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python: Permission denied

    at genericNodeError (node:internal/errors:984:15)
    at wrappedFn (node:internal/errors:538:14)
    at ChildProcess.exithandler (node:child_process:423:12)
    at ChildProcess.emit (node:events:531:35)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:519:28)
    at Pipe.<anonymous> (node:net:339:12) {
  code: 126,
  killed: false,
  signal: null,
  cmd: '/Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python -I /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py'
``` 

I try to run the Python interpreter in my console, and I get permission denied as well. 
```
(teleop) (base) nicolasoulianov@MacBook-Pro-de-Nicolas-2 teleop % /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python
zsh: permission denied: /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python
```

I then change the authorizations of the .venv folder
```
sudo chmod -R 777 .venv
```

I run this command. It doesn't show anything in the command line.
```
 /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python
```

I have this error in the VSCode Python extension logs : failed to parse json
```
2024-12-05 16:59:22.698 [info] > ./teleop/.venv/bin/python -I ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py ~/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py
2024-12-05 16:59:22.701 [error] Failed to parse interpreter information for >> /Users/nicolasoulianov/robots/robots/teleop/.venv/bin/python -I /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/get_output_via_markers.py /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py << with Error: python /Users/nicolasoulianov/.vscode/extensions/ms-python.python-2024.20.0-darwin-arm64/python_files/interpreterInfo.py returned bad JSON () (SyntaxError: Unexpected end of JSON input)
```

🤔

**Comment by zanieb** [comment]
Have you tried updating VSCode to the latest version?

**Comment by oulianov** [comment]
I am on the VSCode latest version available

![Image](https://github.com/user-attachments/assets/c9ac2fc8-2e0c-468e-9467-b9c2eb62172f)

```
Version: 1.95.3 (Universal)
Commit: f1a4fb101478ce6ec82fe9627c43efbf9e98c813
Date: 2024-11-13T14:50:04.152Z (3 wks ago)
Electron: 32.2.1
ElectronBuildId: 10427718
Chromium: 128.0.6613.186
Node.js: 20.18.0
V8: 12.8.374.38-electron.0
OS: Darwin arm64 23.6.0
```

**Comment by zanieb** [comment]
I'm sorry I'm out of suggestions here if you've updated VS Code and restarted it. We can't reproduce this problem.

**Comment by oulianov** [comment]
Hmm... Thank you for the support. 

I assume there must something cursed about my setup :  
- multiple .venv in a single VSCode workspace
- nested python projects

I'll simplify my setup and keep you posted.

In the meanwhile, using an additional .venv for development and uv for deployment works good enough.

Thank you for the support!

**Comment by NellyWhads** [comment]
I'd just like to confirm that I'm using the following setup with success:

- Python monorepo
- Single uv workspace with multiple projects
- using `direnv` to set [`UV_PROJECT_ENVIRONMENT`](https://docs.astral.sh/uv/configuration/environment/?query=preview#uv_project_environment) within each project directory such that `uv` creates a separate `.venv` in the project root instead of the workspace root
- VSCode opened to the workspace directory -> manually selecting the appropriate `.venv` -> `Reload Window`
- Testing integration also works so long as I point VSCode to the specific `pyproject.toml` which houses the `pytest` config for the specific project.

**Comment by cam-at-tactiq** [comment]
I have the same issues as @oulianov, I have a large monorepo and each python project is independent, we don't use the workspaces feature yet.

When selecting the interpreter VS Code follows the symlinks back to `~/.local/share/uv/python/...` without access to any of the libraries I have installed.

**Comment by zanieb** [comment]
Can someone please share a _minimal_ project this reproduces on and a script to setup the environments?

**Comment by chanderson0** [comment]
I can reproduce this locally in my development monorepo. Here's a minimal repro. If there's any other information I can provide, let me know. 

- uv version: 0.5.24
- VSCode version: 1.96.4
- mspython-python extension version: 2024.22.2
- ms-python.vscode-pylance extension version: 2025.1.101

Instructions:
1. Run `./setup.sh`
2. Open VSCode
3. In VSCode, open `projects/project1/hello.py`
4. Run `Python > Select Interpreter`
5. Enter Interpreter Path...
6. Find...
7. Select `path/to/projects/project1/venv/bin/python`
8. Note that Pylance reports `Import "urllib3" could not be resolved`
9. Reload Window
10. Note that Pylance still reports `Import "urllib3" could not be resolved`
11. Rinse and repeat with `project2/hello.py`


`setup.sh`:
```sh
root=$(realpath "$(dirname "$0")")

cd $root
mkdir -p projects/project1
cd projects/project1
uv python find --python-preference only-managed --system
uv init --no-workspace --python-preference only-managed
uv add urllib3
uv sync
cat <<EOF > hello.py
import urllib3
print("p1", urllib3.__file__)
EOF
uv run hello.py
echo ""

cd $root
mkdir -p projects/project2
cd projects/project2
uv python find --python-preference only-system --system
uv init --no-workspace --python-preference only-system
uv add python-dateutil
uv sync
cat <<EOF > hello.py
import dateutil
print("p2", dateutil.__file__)
EOF
uv run hello.py
echo ""
```

Terminal output:
```
/Users/chris/.local/share/uv/python/cpython-3.10.14-macos-aarch64-none/bin/python3.10
Initialized project `project1`
Using CPython 3.10.14
Creating virtual environment at: .venv
Resolved 2 packages in 59ms
Installed 1 package in 3ms
 + urllib3==2.3.0
Resolved 2 packages in 0.20ms
Audited 1 package in 0.01ms
p1 /Users/chris/code/uv-monorepo-vscode/projects/project1/.venv/lib/python3.10/site-packages/urllib3/__init__.py

/Users/chris/.local/share/mise/installs/python/3.10.15/bin/python
Initialized project `project2`
Using CPython 3.10.14
Creating virtual environment at: .venv
Resolved 3 packages in 44ms
Installed 2 packages in 2ms
 + python-dateutil==2.9.0.post0
 + six==1.17.0
Resolved 3 packages in 0.22ms
Audited 2 packages in 0.00ms
p2 /Users/chris/code/uv-monorepo-vscode/projects/project2/.venv/lib/python3.10/site-packages/dateutil/__init__.py
```

**Comment by zanieb** [comment]
@chanderson0 thanks for all the details. Unfortunately, I followed your instructions and could not reproduce the problem.

**Comment by chanderson0** [comment]
Thanks for the quick reply!

After playing with a lot more, I think part of the issue is that there's a difference in (a) using the GUI to select a Python interpreter path (which appears to resolve symlinks and therefore breaks any link to the actual local venv) and entering either (b) an absolute path or (c) a path relative to the workspace root (which appears to not resolve symlinks, and does work as intended).

So maybe the resolution is to always enter an actual path and not the filesystem picker as in (b) and (c). It's working for me, anyway! If I get a chance, I'll open an issue with Pylance, since that's a confusing behavior.

**Comment by zanieb** [comment]
I'm using the GUI to select an interpreter path and was always using an absolute path.

I honestly don't know how to get a filesystem picker to show up :D

**Comment by oulianov** [comment]
Coming back to this issue. Here's a foolproof technique my buddy @Jeerhz showed me to select a uv `.venv` as an active interpreter in vscode. 


1. First, copy the absolute path to the `.venv`. **Not** to the /bin/python, only the .venv folder.

<img width="355" alt="Image" src="https://github.com/user-attachments/assets/2492bd12-aed1-4cbc-8185-6aa33ad368a0" />

2. Do Cmd+Shift+P, then look for "Python: Select interpreter". Press Enter. Select the option: "Enter interpreter path".

<img width="535" alt="Image" src="https://github.com/user-attachments/assets/d9e57cc3-0b3f-4f52-bd55-1750900e4639" />

3. Now in this field you copy paste the absolute path to the .venv folder. 

<img width="456" alt="Image" src="https://github.com/user-attachments/assets/249a183f-df01-4db9-bd62-1343895d80d0" />

<img width="397" alt="Image" src="https://github.com/user-attachments/assets/8071a06f-d38b-426b-ac9d-1da8c8c36128" />

Press enter and now you have the right Python environment with all the dependencies. 


**Still, it would be nice if we didn't have to do all these shenanigans.** In a monorepo, I switch from python project and dependencies all the time. The current procedure is pretty painful, not documented, and not trivial. 

It would be great to have all the `.venv` show up in the interpreter list. Even better, the active interpreter could switch automatically depending on the opened file. 

@zanieb are there any plans to improve the uv experience in vscode?

**Comment by zanieb** [comment]
To be clear, I agree that something must be wrong because we keep hearing reports. I just can't tell what exactly is wrong since each time I try to reproduce, it works fine for me.

**Comment by chanderson0** [comment]
I'm pretty sure that the issue originally highlighted in this thread (and the one I was facing) derives from the VSCode behavior that resolves symlinks when you open the file picker during interpreter selection. That's probably on the VSCode side to fix.

But there may be something additionally happening where uv venvs aren't being picked up automatically, and so someone would need to either (a) add uv support into https://github.com/microsoft/python-environment-tools and https://github.com/microsoft/vscode-python (see e.g. https://github.com/microsoft/vscode-python/pull/24442 and https://github.com/microsoft/python-environment-tools/pull/172 for Pixi doing this recently) or (b) figure out why the existing virtualenv locator isn't working against these environments.

**Comment by zanieb** [comment]
> I'm pretty sure that the issue originally highlighted in this thread (and the one I was facing) derives from the VSCode behavior that resolves symlinks when you open the file picker during interpreter selection. That's probably on the VSCode side to fix.

I'm surprised this hasn't come up before as symbolically linked interpreters in virtual environments are very common on Unix.

We use standards-compliant virtual environments and these _should_ be found by their normal locator since it includes our default name of `.venv`? https://github.com/microsoft/python-environment-tools/blob/1abe5cec5ebfbe97ca71746a4cfc7fe89bddf8e0/crates/pet/src/find.rs#L262

**Comment by NellyWhads** [comment]
To throw another loop in this - our setup is extremely similar and we are able to use the regular VSCode Interpreter selector. We're running the latest VSCode and I've tried it on the latest Cursor. Works stably on MacOS, and on remote development hosts both inside and outside devcontainers with no issues.

**Comment by cloudycotton** [comment]
I encountered this issue just now with my tauri setup. Vscode doesn't recognise the path in .venv created by default.

**Comment by jdumas** [comment]
Hi. I'm also running into this. I have a `.venv` in a `scripts/` subfolder of my vscode project:
```
scripts/.venv/bin/python@ -> /opt/homebrew/Caskroom/miniconda/base/bin/python3
```

Now if I add the interpreter via the vscode UI, it seems to be adding the target path and think I'm in the base conda environment. But if manually specify `<my_repo>/scripts/.venv/bin/python` in in the vscode popup to find the interpreter (i.e. without going through macOS Finder), then it establishes the correct venv.

Seems to be more an issue with vscode than `uv` I would say.

**Comment by cloudycotton** [comment]
I’ve noticed another issue: when I install dependencies and package a script with uvx pyinstaller main.py, the dependencies aren’t bundled as expected. This might be related.

**Comment by jzazo** [comment]
Sorry, I haven't read the full thread, but I will add something in case it helps. I don't experience this issue with VSCode, but whenever my VSCode was misconfigured was because it was not using the workspace properly:
https://code.visualstudio.com/docs/editor/workspaces/workspaces

If you have a single-folder workspace you open your workspace with Open Folder, if you have a monorepo, you select open workspace from file, and you need proper settings.

I have a single-folder repo and everything shows nicely for me using uv.

**Comment by geozeke** [comment]
Just adding my issue to this thread. It seemed like virtual environments were working fine in vscode for a long time, now I experience the same issue with them not activating when opening the terminal. I tried the tips about typing-out the `.venv` path as described above, without luck. I'm managing my environments with `uv`, in separate project folders within a monorepo workspace.

Maybe this is a new bit of information:

I'm using macOS, with `zsh`, OhMyZsh, and the `powerlevel10k` theme (all this worked fine, until just recently). Not sure if others with this issue have the same setup.

**Comment by geozeke** [comment]
It's a little "hacky", but I added this to my `.zshrc` file, and it seems rock-solid whenever you open a new terminal in a vscode project folder, within a monorepo workspace. It assumes that (1) you're running a terminal in vscode, and (2) your virtual environment is located in `./.venv` (the uv default). Maybe a good bandaid until this gets resolved.

```shell
# Activate the python virtual environment if it's present *and* you're running
# the terminal in vscode.
if [ "$TERM_PROGRAM" = "vscode" ]; then
    if [ -d "./.venv" ]; then
        source ./.venv/bin/activate
    fi
fi
```

**Comment by stur86** [comment]
Adding to the experience. Same for me. Virtualenvs created by UV worked fine for me until not long ago. Today I tried one and I got hit with these problems; Pylance doesn't recognise the installed libraries. I made sure to select the correct binary for the virtual environment, this was at the root of the project, there was only one virtual environment, etc. etc. Fails with both Python 3.12 and 3.11. I also tried doing the same thing with Poetry, and with that one it works.

Closest thing that this reminds me of is when some time ago the change to allow for a different format to link editable installs. That did mess up my language server's detection back then. Is it possible something similar has happened here?

**Comment by zanieb** [comment]
@stur86 that'd be if you're using `setuptools` e.g. https://github.com/astral-sh/uv/issues/10946 or https://github.com/astral-sh/uv/issues/1708

**Comment by stur86** [comment]
> [@stur86](https://github.com/stur86) that'd be if you're using `setuptools` e.g. [#10946](https://github.com/astral-sh/uv/issues/10946) or [#1708](https://github.com/astral-sh/uv/issues/1708)

Actually my build backend is hatchling right now. I also tried 

```
[tool.uv]
config-settings = { editable_mode = "compat" }
```

as mentioned in another issue and it does not seem to work, in fact it breaks `uv sync`. Haven't dived deep into why that is yet.

**Comment by SimonBirrell** [comment]
Same for me. Multiple folders within the repo, each with their own pyproject.toml and .venv.

The code works when run, but VS Code doesn't find the .venvs and uses the system Python (as displayed on the bottom line). "Select Python interpreter" doesn't list the .venvs.

**Comment by stur86** [comment]
> Same for me. Multiple folders within the repo, each with their own pyproject.toml and .venv.
> 
> The code works when run, but VS Code doesn't find the .venvs and uses the system Python (as displayed on the bottom line). "Select Python interpreter" doesn't list the .venvs.

My case is worse than this; I do manage to select the virtual environment's Python. It just does nothing to fix this. The virtual environment does have the packages installed, it's just that Pylance specifically doesn't seem to notice.

**Comment by jzazo** [comment]
As I mentioned before, have you correctly set up a single or multi-root workspace?
https://code.visualstudio.com/docs/editor/workspaces/workspaces

**Comment by stur86** [comment]
> As I mentioned before, have you correctly set up a single or multi-root workspace? https://code.visualstudio.com/docs/editor/workspaces/workspaces

The doc mentions you don't need to do anything other than opening a folder with VSCode, so yes I guess. But in the end it did resolve when I created a `.vscode` folder and a settings file inside it, even if essentially empty.

**Comment by tomaszbk** [comment]
I also had my venv in a custom path other than the project's root path, and solved this by running:
`ln -s <custom_venv_location> .venv`
it creates a link to cheat vs code into thinking the venv is there

**Comment by gravityrail** [comment]
Assuming .venv in the project root, this worked for me (when switching from poetry to uv)

```bash
rm -rf .venv
uv sync
```

Then in Cursor:

 * CMD-SHIFT-P
 * Clear workspace interpreter setting
 * Click the python env in the bottom right
 * Click "+ Create virtual environment"
 * Click "Venv"
 * Click to reuse the existing .venv

Then all my dependencies resolved.

**Comment by 0LEL0** [comment]
I found that when I use this plugin "Python Environment Manager (deprecated)" it works well, but when I use "Microsoft Python Environments Extension" it can't find .venv/bin/python.

**Comment by hoangquochung1110** [comment]
Hi guys, do we have a working solution to this issue ? 
Basic features of the IDE are not enabled for some reason if I initial venv using `uv venv` command

![Image](https://github.com/user-attachments/assets/1db1e139-2f19-47ee-b47f-8e4b587d65da)

Many thanks

**Comment by rcgreen99** [comment]
If your pyproject.toml is not in the root of your workspace, you may have to select the interpreter.

1. First, **find the path of your interpreter** by going to the directory where you pyprojectl.toml is and `cd`-ing into `.venv/bin/`. If you `ls` here you should see a file called python, which is your interpreter. You can get the directory path using the command `pwd` to get the current directory.

2. Press cmd + shift + P and type in **Python: Select Interpreter**

2. Select **Enter interpreter path...**

3. Paste in the directory your got using `pwd` from before and `python` to the end of it

**Comment by PacoBahena** [comment]
when you select the interpreter, make sure you pass down the interpreter path manually, because if passing via ui (finder/windows explorer) vscode will set the symlink to uv system python causing the link to fail.

**Comment by soutot** [comment]
> If your pyproject.toml is not in the root of your workspace, you may have to select the interpreter.
> 
> 1. First, **find the path of your interpreter** by going to the directory where you pyprojectl.toml is and `cd`-ing into `.venv/bin/`. If you `ls` here you should see a file called python, which is your interpreter. You can get the directory path using the command `pwd` to get the current directory.
> 2. Press cmd + shift + P and type in **Python: Select Interpreter**
> 3. Select **Enter interpreter path...**
> 4. Paste in the directory your got using `pwd` from before and `python` to the end of it

I tried this and I get `Invalid Python interpreter is selected` error:

![Image](https://github.com/user-attachments/assets/3ab9fc45-0474-45ad-b633-0f628a982f51)

`ls`-ing `/Users/user/Documents/my-repo/py-proj1/.venv/bin` I can see both python and python3.12.

Tried a few different paths: 
/Users/user/Documents/my-repo/py-proj1/.venv/bin
/Users/user/Documents/my-repo/py-proj1/.venv/bin/python
/Users/user/Documents/my-repo/py-proj1/.venv/bin/python3.12

None worked. Any hints?

**Comment by NellyWhads** [comment]
Have you tried:
1. Using the pre-release version of the Python extension? This used to fix things for me (but is no longer required for the last ~4 months)
2. Re-loading the VSCode Window, Re-starting the python language server, and/or re-starting the extensions provider?

**Comment by codethief** [comment]
Just to add another anecdotal data point to the discussion: My workspace is a monorepo, whose subfolders each contain a separat uv project / `pyproject.toml` & `.venv` folder. Now I just installed the [teticio.python-envy](https://github.com/teticio/python-envy) extension, which automatically recognizes `.venv` folders and sets the interpreter path depending on the file I am looking at. So far, everything (IntelliSense, type hints, etc.) seems to be working fine!

I'm running
- VSCode 1.98.2
- uv 0.6.9
- ms-python.python 2025.4.0
- ms-python.vscode-pylance 2025.4.1
- teticio.python-envy 0.1.8

**Comment by zcuric** [comment]
I had this issue when using Docker and Django. I notice that `.venv` was messed up with from the Docker so I made it work like this:

Create a separate virtual environment (to differentiate it from the one used in Docker):
```
uv venv .venv-editor
```

Activate the virtual environment:
```
source .venv-editor/bin/activate
```

Install dependencies into the active environment:
```
uv sync --active
```

In VSCode, open the command palette (Cmd+Shift+P), search for “Python: Select Interpreter”, and select the .venv-editor environment (it should point to .venv-editor/bin/python).

**Comment by ben-xD** [comment]
The workarounds mentioned above didn't really help me. `uv run python` or just `./venv/bin/python` was killed with error code 137.

This was happening to me on python 3.11. A quick fix for me was to upgrade the project I was testing to Python 3.12 or 3.13.

**Comment by rayterrill** [comment]
I fixed this by setting the following setting in my user settings.json:
```
    "python.defaultInterpreterPath": "./.venv",
```

**Comment by Victsz** [comment]
My case is the language server never work for pylance when using uv and select python in  ./.venv
![Image](https://github.com/user-attachments/assets/849e23eb-4e84-451a-83ed-aa2c32a52ae4)
But sometimes jedi can work but not always...

![Image](https://github.com/user-attachments/assets/0f14a5ec-5f78-49e3-b88c-44774221be83)

**Comment by jbohnslav** [comment]
This is semi-related, but in case it helps people: I kept getting sigkills as well. I had to 

`rm -rf /Users/USERNAME/.local/share/uv`, then `uv sync` and it all worked.

**Comment by Robinsstudio** [comment]
I also have this issue and my `.venv` is not at the root of the project as well. The workaround fixes it but it's actually not really necessary on my end as using the absolute path to the Python interpreter will work. I just have to choose "Enter interpreter path" and then paste the absolute path to it. It works with the absolute path to `.venv`, `.venv/bin` and even `.venv/bin/python`.

I also figured that it works with paths relative to the root of the workspace. For example, if I have my workspace at `/Users/robinsstudio/my_amazing_project` and my Python interpreter at `/Users/robinsstudio/my_amazing_project/path/to/python/.venv/bin/python`, then I can simply paste `./path/to/python/.venv/bin/python` at the "Enter interpreter path" step.

I hope this will help someone!

**Comment by egorvavilov** [comment]
I encountered the same issue when setting up a multi-repo workspace using uv on a Mac M1, and none of the suggested solutions worked for me initially—especially in getting `VSCode + Pylance + Python + uv` to cooperate.

After some searching, I found a workaround that resolved the issue. Here's what worked:

Project structure:
```md
.
├── child_repo_A
│   ├── .venv
│   ├── pyproject.toml
│   └── ...
├── child_repo_B
│   ├── .venv
│   ├── pyproject.toml
│   └── ...
└── README.md
```

Steps:
1. Open the Command Palette `(Cmd+Shift+P / Ctrl+Shift+P)` in VSCode and select: `Workspaces: Add Folder to Workspace...`
    Add each child repo as a separate folder.
2. Once added, VSCode will treat each child repository as an independent project in the File Explorer.
3. Inside each child repo, create a .vscode/settings.json file with the following content:
```json
{
  "python.defaultInterpretorPath": ".venv/bin/python"
}
```
4. This enables proper interpreter resolution, and Pylance should now work correctly with your environment.

Let me know if this helps or if you run into any edge cases

**Comment by BlackDark** [comment]
I also did not have completions for my local projects but when I did a uv sync  --no-editable i got the completions back working because the folder gets copied to the .venv directory I think. But of course now also the stuff is fixed and not symlinked so changes get not reflected back to the .venv.

Not sure if this is related.

**Comment by konstin** [comment]
I asked upstream (VS Code Python) for input on the matter: https://github.com/microsoft/vscode-python/issues/25068

**Comment by konstin** [comment]
Could you check whether you have the same problem with https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs? From a VS Code maintainer in https://github.com/microsoft/vscode-python/issues/25068:

> Install that extension and reload VS Code. That will cause python extension to replace its features with the new extension. That will soon be replacing all the environment features in python extension.

**Comment by wnowicki** [comment]
I've tried, installed and it didn't help.

Please see as well https://github.com/microsoft/pylance-release/issues/7209

**Comment by zhuoqun-chen** [comment]
Hi all and @oulianov, would you mind trying to add this line in your project's `.vscode/settings.json`:

```json
"python.defaultInterpreterPath": "${workspaceFolder}/.venv"
```

I'm assuming `uv` is installed globally inside your `$PATH`, eg, I'm using `uv` installed with `brew` on my mac.

If getting no luck, also setting up in the UI interface like this:

<img width="670" alt="Image" src="https://github.com/user-attachments/assets/d2a2aba9-820d-4aba-a002-8363f22e112e" />

Using this always gives me the correct IDE features whether in `cursor` or `vscode`. Anyway, this is just my own config so I'm posting here for reference. Hope this can help you.

**Comment by bholagabbar** [comment]
I spent like 3 days trying to fix this uv/autocomplete issue but in vain. This is a waste of time. Great library, terrible support for existing tooling. just CANNOT get it to work in a monorepo setup with pylance and the existing other tools. I've tried everything under the sun, nope

**Comment by codethief** [comment]
@bholagabbar Did you try https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995 ?

**Comment by zanieb** [comment]
@bholagabbar please go open an issue in VS Code or comment on [this existing one](https://github.com/microsoft/vscode-python/issues/25068) instead. We are just using standards-compliant Python virtual environments. There's nothing we can do to fix this behavior — and if there somehow is I'd be happy to hear it.

**Comment by bholagabbar** [comment]
@codethief I tried installing the package, didn't work for me :(

Hey @zanieb , appreciate your reply. Let me layout the step my step commands to replicate this. I really want to help your team (and myself) get to the bottom of this issue and this really shouldn't be something that stops users from using uv. So here goes

I use [Cursor](https://www.cursor.com/)

* `uv init infra --bare` (This acts sort of as the base of my monorepo`
* `cd infra`
* `uv venv`. Then I set the interpreter to this path and restart my terminal too
* `uv init dataos --lib` (Core library to by shared by services`
* `uv init actor-rep-runner` (Service that uses code from dataos` 
* The pyproject.toml files for `infra` and `actor-rep-runner`  are as follows
    ```
    [project]
    name = "infra"
    version = "0.1.0"
    requires-python = ">=3.13"
    dependencies = []
    
    [tool.uv.workspace]
    members = [
        "dataos",
        "actor-rep-runner",
    ]
    ```

    ```
    [project]
    name = "actor-rep-runner"
    version = "0.1.0"
    description = "Add your description here"
    readme = "README.md"
    requires-python = ">=3.13"
    dependencies = [
        "dataos",
    ]
    
    [tool.uv.sources]
    dataos = { workspace = true }
    ```
* I do a `uv sync` from the root `infra/` 
* Create a file inside `dataos` that you want to reuse in your service
  * `cd dataos/src/dataos`
  * `touch hunter.py`
  ```
  def hunt():
      print ("Hunting...")
  ```
  
* `cd ../../../actor-rep-runner/`
This is how my `main.py` looks, with no imports

![Image](https://github.com/user-attachments/assets/5df5bdb4-e42d-4fee-88d4-e309c161df3f)


As you can see, `from dataos.hunter import hunt` is unable to be detected. Any ideas ?

**Comment by codethief** [comment]
@bholagabbar 

> @codethief I tried installing the package, didn't work for me :(

That's unfortunate! What exactly didn't work, though? Selecting the right .venv? Or import resolution as in your screenshot?

Regarding your screenshot: What exactly is the error message when you hover over the import? Are you using Pylance? If so, you might want to look at https://github.com/microsoft/pylance-release/issues/7151 and https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#unresolved-import-warnings .

**Comment by xflcx1991** [comment]
> ❌ The solution mentioned in [#9637 (comment)](https://github.com/astral-sh/uv/issues/9637#issuecomment-2654030261) will not work on all setup.❌ 中 [#9637 (comment)](https://github.com/astral-sh/uv/issues/9637#issuecomment-2654030261) 提到的解决方案不适用于所有设置。
> 
> > if you have a monorepo, you select open workspace from file如果您有 monorepo，请选择 Open workspace from file
> 
> If you’re working in a monorepo, you can select “Open Workspace from File” to open a specific subfolder as its own workspace.如果你在 monorepo 中工作，你可以选择 “Open Workspace from File” 来打开一个特定的子文件夹作为它自己的工作区。
> 
> When you do this, external libraries will still resolve correctly. However, internal imports will only work if they are relative to the subfolder you’re opening as the workspace.执行此作时，外部库仍将正确解析。但是，仅当内部导入相对于您作为工作区打开的子文件夹时，内部导入才有效。
> 
> For example, with the following structure:例如，具有以下结构：
> 
> ```
> main_folder/
> ├── sub_folder/  (opened as workspace)
> │   ├── .venv/
> │   ├── pyfile_1.py
> │   └── pyfile_2.py
> ├── ...
> ```
> 
> Inside `pyfile_2.py`:`pyfile_2.py` 内幕：
> 
> ```
> from pyfile_1 import whatever  # ❌ Won't work
> from typing import Any         # ✅ Works (standard library)
> ```
> 
> So, imports like from `pyfile_1` import ... won’t resolve unless `pyfile_1.py` is discoverable from the workspace root — in this case, sub_folder.因此，像 from `pyfile_1` import 这样的 imports ...除非可以从工作区根（在本例中为 sub_folder）发现 `pyfile_1.py`，否则不会解析。
> 
> ✅ **The solution mentioned in [this comment](https://github.com/astral-sh/uv/issues/9637#issuecomment-2626203937) worked for me** with this setup:✅ **[ 此评论](https://github.com/astral-sh/uv/issues/9637#issuecomment-2626203937)中提到的解决方案适用于此设置 **：
> 
> ```
> main_folder/ (opened as workspace)
> ├── sub_folder/  
> │   ├── .venv/
> │   ├── pyfile_1.py
> │   └── pyfile_2.py
> ├── ...
> ```
> 
> That said, it didn’t work immediately — possibly due to caching 🤷 or something similar. Restarting / closing the folder or else might help.也就是说，它并没有立即工作——可能是由于缓存🤷或类似原因。重新启动/关闭文件夹或其他方式可能会有所帮助。

Thank you very much, it turned out to be this "Open Workspace from File". I'm also mono-project, I was inspired, I chose "Add Fold to workspace", I chose the root of the python project folder created by uv, and it worked, now cursor workspace has two entries, one is the mono-project root, and the other is the python project root that I just added, Now the python code has selected the interpreter path to be the one in the .venv directory, and the import statement is normal, and there is no warning that the package cannot be found.

**Comment by bholagabbar** [comment]
Thanks! Was finally able to get it to work. One common thread is that the code editor takes time to recognize the imports and needs to be restarted. I'm not really sure what worked, but I documented everything in my README as such:

```
* `uv venv` - Creates a `.venv/` virtual environment folder
* `source .venv/bin/activate` - Activates the virtual environment
* `uv sync` - Installs dependencies for all projects in the workspace
* Ensure you have a `.env` wherever we have a `.env.sample` file
* [Select interpreter](https://github.com/astral-sh/uv/issues/9637#issuecomment-2626203937)
* Install [python-envy](https://github.com/teticio/python-envy) extension for vscode
* Restart your terminal and window for autocompletion to work

`uv` is tricky to setup, here are some selected github issues that address common annoyances: [1](https://github.com/astral-sh/uv/issues/9637#issue-2717716303), [2](https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995)

```

**Comment by bholagabbar** [comment]
@zanieb I have a (somewhat) related question. Why does `uv` not treat workspace members like automatic dependencies ? 

Here's an example
```
[project]
name = "infra"
version = "0.1.0"
requires-python = ">=3.13"

[tool.uv.workspace]
members = [
    "dataos",
    "actor-rep-runner",
]
```
This is the root `pyproject.toml` for the workspace. If I do `uv sync` here, it actually wipes all the existing dependencies it had installed, when i did `uv sync` from WITHIN the `dataos/` and `actor-rep-runner/` folders

What fixed it was adding these 2 blocks

```
dependencies = [
    "dataos",
    "actor-rep-runner",
]

[tool.uv.sources]
dataos = { workspace = true }
actor-rep-runner = { workspace = true }
```

I really shouldn't have to add the above ? Seems obvious to me personally, or maybe I am opinionated 🤷‍♂

**Comment by charliermarsh** [comment]
@bholagabbar -- Because workspaces consist of multiple members, and each member can depend on any subset (or none, or all) of the other members. For example, you might not want `dataos` to depend on `actos-rep-runner`, right? The workspace root is no different -- it's just another member, and each member needs to declare its own dependencies.

If you have further questions, I suggest joining the Discord? (Since this issue in particular has a lot of subscribers, and it's better not to ping them all for unrelated queries.)

**Comment by pdpark-wex** [comment]
The only thing that worked for me was reverting to the older "js" python locator with this setting: "python.locator": "js" - from [here](https://stackoverflow.com/questions/78886125/vscode-python-extension-loading-forever-saying-reactivating-terminals#answer-78886126). Obviously not a great long term solution, but nothing else worked.

I reported this [here](https://github.com/microsoft/vscode-python/issues/25120#issuecomment-2923529911)

**Comment by konstin** [comment]
@pdpark-wex Does this mean you also tried https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs and that didn't work?

**Comment by pdpark-wex** [comment]
> [@pdpark-wex](https://github.com/pdpark-wex) Does this mean you also tried https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs and that didn't work?

I have not tried it yet.

**Comment by 7l-hyuk** [comment]
Anyone using WSL should check the following:

* In vscode, press `ctrl + shift + P`
* Select `WSL: Connect to WSL`

VSCode will now connect to your WSL instance and automatically detect your virtual environment.

**Comment by thefabus** [comment]
I'm not sure if this has been mentioned yet, but I had the same issue (Select Interpreter appears to do nothing), and what worked for me was to disable managed python.

The smoking gun was the managed interpreter crashing, however, after uninstalling all managed python interpreters I've been unable to reproduce this. The following only occurred in iTerm, in the VSCode terminal it would just hang:

```
❯ uv python install --reinstall
Installed Python 3.11.13 in 1.52s
 + cpython-3.11.13-macos-aarch64-none
❯ deactivate
❯ source .venv/bin/activate
❯ python --version
zsh: killed     python --version
```

I uninstalled all managed python versions:
```
❯ uv python list
cpython-3.14.0b2-macos-aarch64-none                 <download available>
cpython-3.14.0b2+freethreaded-macos-aarch64-none    <download available>
cpython-3.13.5-macos-aarch64-none                   /opt/homebrew/bin/python3.13 -> ../Cellar/python@3.13/3.13.5/bin/python3.13
cpython-3.13.5-macos-aarch64-none                   /opt/homebrew/bin/python3 -> ../Cellar/python@3.13/3.13.5/bin/python3
cpython-3.13.5-macos-aarch64-none                   <download available>
cpython-3.13.5+freethreaded-macos-aarch64-none      <download available>
cpython-3.12.11-macos-aarch64-none                  <download available>
cpython-3.11.13-macos-aarch64-none                  <download available>
cpython-3.10.18-macos-aarch64-none                  <download available>
cpython-3.9.23-macos-aarch64-none                   <download available>
cpython-3.9.6-macos-aarch64-none                    /usr/bin/python3
cpython-3.8.20-macos-aarch64-none                   <download available>
pypy-3.11.11-macos-aarch64-none                     <download available>
pypy-3.10.16-macos-aarch64-none                     <download available>
pypy-3.9.19-macos-aarch64-none                      <download available>
pypy-3.8.16-macos-aarch64-none                      <download available>
graalpy-3.11.0-macos-aarch64-none                   <download available>
graalpy-3.10.0-macos-aarch64-none                   <download available>
graalpy-3.8.5-macos-aarch64-none                    <download available>
```

Then reinitializing with system python fixed it for me:
```
❯ rm -rf .venv
❯ uv venv --no-managed-python
Using CPython 3.13.5 interpreter at: /opt/homebrew/opt/python@3.13/bin/python3.13
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
```

Hope this helps.

**Comment by bholagabbar** [comment]
This issue seems to be initiation ritual into uv lol. Happy user now, but the 3 days I spent trying to figure this out were ridiculous

Sent via Superhuman iOS ( ***@***.*** )

**Comment by frodera** [comment]
This solution from @thefabus is the only one that has worked for me.

> ❯ rm -rf .venv
❯ uv venv --no-managed-python
Using CPython 3.13.5 interpreter at: /opt/homebrew/opt/python@3.13/bin/python3.13
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate

The workaround of "Selecting .venv as an active interpreter" suggested did not do the trick for me for some reason. 🤷🏻

**Comment by konstin** [comment]
From the other thread, it seems that the preview environment extension also fixes this (https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs, https://github.com/microsoft/vscode-python/issues/25068#issuecomment-2967309800)

**Comment by lmanc** [comment]
@konstin, the locator fix suggested by @pdpark-wex worked for me, too.

I tried everything else, the `.venv` absolute path and the new Python Environments extension, but the only thing that fixed it was switching back to the old locator.

**Comment by day0n** [comment]
+1

**Comment by Duc-NSH** [comment]
> The only thing that worked for me was reverting to the older "js" python locator with this setting: `"python.locator": "js"` - from [here](https://stackoverflow.com/questions/78886125/vscode-python-extension-loading-forever-saying-reactivating-terminals#answer-78886126). Obviously not a great long term solution, but nothing else worked.
> 
> I reported this [here](https://github.com/microsoft/vscode-python/issues/25120#issuecomment-2923529911)

Thank you very much! This solution worked for me. Although I mostly use NeoVim, I still need to use VsCode for certain tasks. The bug where VsCode doesn't detect non-root .venv makes me dislike opening it every time.

**Comment by hsiehbocheng** [comment]
Not sure if I’m misunderstanding your issue, but I ran into something very similar.
VS Code couldn’t find the Python interpreter in my .venv from uv, and I eventually discovered it was because the virtual-env lived inside a sub-folder (e.g. project/sub-module/.venv).

By default a VS Code workspace only scans the top-level folder for interpreters, so anything in deeper paths is ignored. The fix was to switch to a multi-root workspace (File ▸ Add Folder to Workspace…) and add the sub-folder as its own root. As soon as I did that, the .venv interpreter showed up.

Docs for reference: https://code.visualstudio.com/docs/editing/workspaces/multi-root-workspaces

**Comment by codethief** [comment]
@Duc-NSH @hsiehbocheng 

> The bug where VsCode doesn't detect non-root .venv

> I eventually discovered it was because the virtual-env lived inside a sub-folder (e.g. project/sub-module/.venv).
> By default a VS Code workspace only scans the top-level folder for interpreters, so anything in deeper paths is ignored.

Correct, and – with all due respect – we have discussed this various times already (see e.g. the [4th comment here](https://github.com/astral-sh/uv/issues/9637#issuecomment-2518779669)). The bug/deficiency is in Pyright/Pylance[0], and has nothing to do with uv.

Overall, I feel like we're going in circles, so **let me summarize all the solutions & workarounds people here in this thread have identified so far**. I added a suffix `(?)` to solutions I haven't verified myself and where there hasn't been a lot of feedback yet.

**Solution 1:**  Set the interpreter path by hand manually, as discussed [in the very first post](https://github.com/astral-sh/uv/issues/9637#issue-2717716303) here as well as [various](https://github.com/astral-sh/uv/issues/9637#issuecomment-2552026597) [others](https://github.com/astral-sh/uv/issues/9637#issuecomment-2756287897). Note that there seems to be an additional issue regarding symlinks, see [1](https://github.com/astral-sh/uv/issues/9637#issuecomment-2619993750), [2](https://github.com/astral-sh/uv/issues/9637#issuecomment-2773598069), [3](https://github.com/astral-sh/uv/issues/9637#issuecomment-2856524057). Either way, this solution doesn't work great if you're in a monorepo or often switch between multiple projects/workspaces in VSCode.

**Solution 2:** Use a multi-root workspace in VSCode, i.e. one root for every Python project. This isn't exactly great if you're in a monorepo with a lot of projects and collaborators, each of whom will have to set up their VSCode instance by hand, but it seems it is what the [Pyright devs recommend](https://github.com/microsoft/pyright/discussions/7979#discussioncomment-9527473).

**Solution 3**: Like 1 but use an extension like [teticio.python-envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy) to automatically set the interpreter path based on the file you're currently looking at, compare https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995 . This is what I currently use and it has been working quite well in my monorepo setup.

**Solution 4(?)**: Use the new [ms-python.vscode-python-envs](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs) extension, as per https://github.com/astral-sh/uv/issues/9637#issuecomment-2979632654 .

**Solution 5(?)**: Set `"python.locator": "js"`, compare https://github.com/astral-sh/uv/issues/9637#issuecomment-2923535889 .

**Side note regarding solutions 1 and 3:** You might also want to add a `pyrightconfig.json` to your workspace root and set different [`executionEnvironments`](https://microsoft.github.io/pyright/#/configuration?id=execution-environment-options) for your different projects there. This should ensure that import suggestions work as expected. E.g.:

```javascript
{
    "executionEnvironments": [
        {
            "root": "./python-project1"
        },
        {
            "root": "./python-project2"
        },
        // …
    ],
    // …
}
```


[0]: Related tickets: 
- https://github.com/microsoft/pyright/issues/10498
- https://github.com/microsoft/pylance-release/issues/7151
- https://github.com/microsoft/pylance-release/issues/3899
- https://github.com/microsoft/pylance-release/issues/5564
- https://github.com/microsoft/vscode-python/issues/21204

**Comment by jyaros** [comment]
Personally i was able to add uvenv as an interpreter only once I reopened the git repo at the top level of the workspace so that the env folder was at root. If you dont **need** to open up a higher parent directory for your work, this is the easiest way to get around the issue ...

**Comment by zanieb** [comment]
If there's a canonical way to configure this correctly, e.g., as described by @codethief, I think we should add it to an integration guide in the documentation.

**Comment by ra9fael** [comment]
> The only thing that worked for me was reverting to the older "js" python locator with this setting: `"python.locator": "js"` - from [here](https://stackoverflow.com/questions/78886125/vscode-python-extension-loading-forever-saying-reactivating-terminals#answer-78886126). Obviously not a great long term solution, but nothing else worked.
> 
> I reported this [here](https://github.com/microsoft/vscode-python/issues/25120#issuecomment-2923529911)

Thank you, work like a charm.

**Comment by eleanorjboyd** [comment]
Hello everyone! Wanted to hop in per @konstin suggestion. I am an engineer on the VS Code team specifically working on Python and environments. We are beginning to roll out a new extension called the Python Environments Extension which is a new extension dedicated to all environment and package management for Python in VS Code. This will serve as the backbone for Python support in VS Code, a required extension which all Python extensions (the core Python, the Python Debugger, and various linter/formatter extensions) will rely on.

With this being said, we encourage everyone encountering issues with their environments in VS Code to install the new extension and add to their user settings:  `python.usePythonEnvironmentsExtension: true`. Once enabled you may find your issue is fixed (as we have made significant improvements to the environment experience) or it may still persist. If you still have it, please file an issue on [this repo](https://github.com/microsoft/vscode-python-environments) and I will get back to you to figure out and solve the bug. This will make sure your bug is fixed moving forward instead of on the soon-to-be deprecated code for environments which lives on the Python extension. 

Feel free to reach out with any outstanding questions in that repository and I hope everyone adopts the new extension as we are excited about the new features which it enables!

**Comment by ggirelli** [comment]
> With this being said, we encourage everyone encountering issues with their environments in VS Code to install the new extension and add to their user settings: `python.usePythonEnvironmentsExtension: true`.

This fixed the issue for me. Thank you so much! 🙏🏻

**Comment by hsuabina** [comment]
>Solution 3: Like 1 but use an extension like [teticio.python-envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy) to automatically set the interpreter path based on the file you're currently looking at, compare https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995 . This is what I currently use and it has been working quite well in my monorepo setup.

I'm also working on a monorepo (with packages of all sorts) and for someone that's not so used to work on python projects this solution was simple and effective.

**Comment by codethief** [comment]
@eleanorjboyd I've seen your new extension be mentioned a few times already but so far no one ever spelled out what it does differently compared to the old Python extension. Could you elaborate? Does it support monorepos? (I.e. multiple Python projects / venvs per workspace.) Does it automatically recognize Python projects / venvs or do I have to configure things by hand? Etc.

**Comment by matthias-busch** [comment]
> Solution 3: Like 1 but use an extension like [teticio.python-envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy) to automatically set the interpreter path based on the file you're currently looking at, compare https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995 . This is what I currently use and it has been working quite well in my monorepo setup.

Gotta say this does work quite well in my monorepos. Thanks for the tip!

**Comment by mironnn** [comment]
had the same issue.

Disabling this buggy sh..t solved the issue.
https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs is recommended to be uninstalled.

**Comment by eleanorjboyd** [comment]
Hi @mironnn- would appreciate if you file an issue on our repo if you are seeing a problem so I can fix it! Thanks

**Comment by zcuric** [comment]
> Hello everyone! Wanted to hop in per [@konstin](https://github.com/konstin) suggestion. I am an engineer on the VS Code team specifically working on Python and environments. We are beginning to roll out a new extension called the Python Environments Extension which is a new extension dedicated to all environment and package management for Python in VS Code. This will serve as the backbone for Python support in VS Code, a required extension which all Python extensions (the core Python, the Python Debugger, and various linter/formatter extensions) will rely on.
> 
> With this being said, we encourage everyone encountering issues with their environments in VS Code to install the new extension and add to their user settings: `python.usePythonEnvironmentsExtension: true`. Once enabled you may find your issue is fixed (as we have made significant improvements to the environment experience) or it may still persist. If you still have it, please file an issue on [this repo](https://github.com/microsoft/vscode-python-environments?rgh-link-date=2025-08-18T16%3A02%3A20.000Z) and I will get back to you to figure out and solve the bug. This will make sure your bug is fixed moving forward instead of on the soon-to-be deprecated code for environments which lives on the Python extension.
> 
> Feel free to reach out with any outstanding questions in that repository and I hope everyone adopts the new extension as we are excited about the new features which it enables!

Does this apply in Cursor as well? 🙈

**Comment by ksaurabh21** [comment]
I am using VS code on Win11 VM on azure. I have created a .venv using uv but none of the solutions suggested above worked for me. I am trying to run a notebook (have jupyter extension and ipykernel installed) but unable to have VS code detect the venv.

**Comment by piehld** [comment]
> With this being said, we encourage everyone encountering issues with their environments in VS Code to install the new extension and add to their user settings: `python.usePythonEnvironmentsExtension: true`. 

Is it `python.usePythonEnvironmentsExtension` or `python.useEnvironmentsExtension`? The extension docs suggest the latter, and the former is not showing up as a valid setting.

<img width="870" height="421" alt="Image" src="https://github.com/user-attachments/assets/0cf45e70-0a27-4a40-92d8-1be5cbca5cc8" />

**Comment by djsamseng** [comment]
In my case, **not** selecting "Select at workspace level" fixed the issue.

My setup was:
- A global uv venv located outside of the VSCode workspace (ex: ~/.venv)
- A VSCode workspace with multiple projects
- Python: Select interpreter -> Select at workspace level -> Enter interpreter path -> intellisense does not work, installed modules (via uv pip install) are not found

What did work was:
Python: Select interpreter -> ***Project Name*** -> Enter interpreter path -> intellisense works

**Comment by akk29** [comment]
Resolved, I just selected path of interpreter in from VS code bottom right corner in .venv/Scripts/python, after that it got resolved, though I have also uninstalled the extension mentioned above , added that config also didn't work that either (reverted it). Not sure if uninstalling extension have any effect on this. Path selection for interpreter have definitely fixed this issue.

**Comment by juhaszp95** [comment]
Just leaving a comment here to say that https://github.com/astral-sh/uv/issues/9637#issuecomment-2813097995 works for me, and is the only solution which is working for me at the moment (i.e. the new Python Environments extension didn't solve the issue for me).

**Comment by tinodj** [comment]
I hit this when using uv, and actually the problem is finder resolving symlink on MacOS, described here: https://github.com/astral-sh/uv/issues/8706#issuecomment-2773598984

With python.usePythonEnvironmentsExtension: true this is not fixed, and this extension even it makes impossible to enter path manually, only to select it from finder, which then resolves to original uv python. However, extension is very nice itself and there are lot of useful features.

**Comment by cs-mshah** [comment]
> In my case, **not** selecting "Select at workspace level" fixed the issue.
> 
> My setup was:
> 
> * A global uv venv located outside of the VSCode workspace (ex: ~/.venv)
> * A VSCode workspace with multiple projects
> * Python: Select interpreter -> Select at workspace level -> Enter interpreter path -> intellisense does not work, installed modules (via uv pip install) are not found
> 
> What did work was: Python: Select interpreter -> _**Project Name**_ -> Enter interpreter path -> intellisense works

This is what works for me as well. Thanks!

**Comment by ericbmoreira** [comment]
> had the same issue.
> 
> Disabling this buggy sh..t solved the issue. https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs is recommended to be uninstalled.

This was the solution for me! I could run Shift + Enter using conda but uv venv would not work. I would keep getting "Python interpreter not found..." When I disabled that extension it worked as expected. Thanks for the tip!

**Comment by tobsecret** [comment]
>     had the same issue.
> 
>     Disabling this buggy sh..t solved the issue. https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-python-envs is recommended to be uninstalled.

This also solved the issue for me. Now it loads instantly.

**Comment by konstin** [comment]
I'm going to close this, the solution seems to be working and I've updated the original post with the solution. If the problem persist, please open a new issue with a detailed description of what you did and what didn't work.
